src/server/twemcache-rs/src/admin.rs

```diff
@@ -0,0 +1,413 @@
+// Copyright 2021 Twitter, Inc.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+use crate::event_loop::EventLoop;
+use crate::protocol::admin::*;
+use crate::session::*;
+use crate::*;
+use crossbeam_channel::Receiver;
+use crossbeam_channel::Sender;
+use metrics::Stat;
+
+use rustcommon_fastmetrics::{Metric, Source};
+
+use boring::ssl::{HandshakeError, MidHandshakeSslStream, Ssl, SslContext, SslStream};
+use mio::event::Event;
+use mio::net::TcpListener;
+use strum::IntoEnumIterator;
+
+use std::convert::TryInto;
+use std::io::{Error, ErrorKind};
+
+/// A `Admin` is used to bind to a given socket address and handle out-of-band
+/// admin requests.
+pub struct Admin {
+    addr: SocketAddr,
+    config: Arc<Config>,
+    listener: TcpListener,
+    poll: Poll,
+    ssl_context: Option<SslContext>,
+    sessions: Slab<Session>,
+    message_receiver: Receiver<Message>,
+    message_sender: Sender<Message>,
+}
+
+pub const LISTENER_TOKEN: usize = usize::MAX;
+
+impl Admin {
+    /// Creates a new `Admin` event loop.
+    pub fn new(config: Arc<Config>) -> Result<Self, Error> {
+        let addr = config.admin().socket_addr().map_err(|e| {
+            error!("{}", e);
+            std::io::Error::new(std::io::ErrorKind::Other, "Bad listen address")
+        })?;
+        let mut listener = TcpListener::bind(addr).map_err(|e| {
+            error!("{}", e);
+            std::io::Error::new(std::io::ErrorKind::Other, "Failed to start tcp listener")
+        })?;
+        let poll = Poll::new().map_err(|e| {
+            error!("{}", e);
+            std::io::Error::new(std::io::ErrorKind::Other, "Failed to create epoll instance")
+        })?;
+
+        // TODO(bmartin): does this need to be configurable to require TLS for
```

Comment:
This is probably worth addressing before merge.

Reply:
We should eventually move things up popping off the top level src directory. This repo should be organized as a workspace of crates. Eg: `server/twemcache`, `server/redis`, `storage/segcache`. Not sure if you have thoughts on how to handle this in the current hybrid state and/or if we want to retain cmake builds for rust backends.